File: sort-array-by-parity-ii/solution922.go
// https://leetcode.com/problems/sort-array-by-parity-ii/

package main

import "fmt"

// T: O(n), S: O(n)
func sortArrayByParityII(nums []int) []int {
	n := len(nums)
	j, k := 0, 1
	ans := make([]int, n)
	for _, num := range nums {
		if num&1 == 0 {
			ans[j] = num
			j += 2
		} else {
			ans[k] = num
			k += 2
		}
	}
	return ans
}

// T: O(n), S: O(1)
func sortArrayByParityIIOpt(nums []int) []int {
	n := len(nums)
	evenPtr, oddPtr := 0, 1
	for evenPtr < n && oddPtr < n {
		if nums[evenPtr]&1 == 0 {
			evenPtr += 2
		} else if nums[oddPtr]&1 == 1 {
			oddPtr += 2
		} else {
			nums[evenPtr], nums[oddPtr] = nums[oddPtr], nums[evenPtr]
		}
	}
	return nums
}

func main() {
	nums1 := []int{4, 2, 5, 7}
	nums2 := []int{2, 3}
	fmt.Println(sortArrayByParityII(nums1))    // [4, 5, 2, 7]
	fmt.Println(sortArrayByParityIIOpt(nums1)) // [4, 5, 2, 7]
	fmt.Println(sortArrayByParityII(nums2))    // [2, 3]
	fmt.Println(sortArrayByParityIIOpt(nums2)) // [2, 3]
}

File: sort-array-by-parity-ii/Solution922.java
// https://leetcode.com/problems/sort-array-by-parity-ii/

import java.util.Arrays;

class Solution922 {

    // T: O(n), S: O(n)
    public int[] sortArrayByParityII(int[] nums) {
        int n = nums.length;
        var ans = new int[n];
        int j = 0, k = 1;
        for (int i = 0; i < n; i++) {
            if ((nums[i] & 1) == 0) {
                ans[j] = nums[i];
                j += 2;
            } else {
                ans[k] = nums[i];
                k += 2;
            }
        }
        return ans;
    }

    // T: O(n), S: O(1)
    public int[] sortArrayByParityIIOpt(int[] nums) {
        int n = nums.length;
        int evenPtr = 0, oddPtr = 1;
        while (evenPtr < n && oddPtr < n) {
            if ((nums[evenPtr] & 1) == 0) {
                evenPtr += 2;
            } else if ((nums[oddPtr] & 1) == 1) {
                oddPtr += 2;
            } else {
                int temp = nums[evenPtr];
                nums[evenPtr] = nums[oddPtr];
                nums[oddPtr] = temp;
            }
        }
        return nums;
    }

    public static void main(String[] args) {
        var sol = new Solution922();
        var nums1 = new int[] { 4, 2, 5, 7 };
        var nums2 = new int[] { 2, 3 };
        System.out.println(Arrays.toString(sol.sortArrayByParityII(nums1))); // [4,5,2,7]
        System.out.println(Arrays.toString(sol.sortArrayByParityIIOpt(nums1))); // [4,5,2,7]
        System.out.println(Arrays.toString(sol.sortArrayByParityII(nums2))); // [2,3]
        System.out.println(Arrays.toString(sol.sortArrayByParityIIOpt(nums2))); // [2,3]
    }

}
